src/Channels.Networking.Sockets/SocketConnection.cs

```diff
@@ -0,0 +1,240 @@
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Channels.Networking.Sockets
+{
+    public class SocketConnection : IDisposable
+    {
+        private Socket socket;
+        private Channel input, output;
+
+        public IReadableChannel Input => input;
+        public IWritableChannel Output => output;
+        private ChannelFactory ownedChannelFactory;
+
+        internal SocketConnection(Socket socket, ChannelFactory channelFactory)
+        {
+            socket.NoDelay = true;
+            this.socket = socket;
+            if(channelFactory == null)
+            {
+                ownedChannelFactory = channelFactory = new ChannelFactory();
+            }
+            input = channelFactory.CreateChannel();
+            output = channelFactory.CreateChannel();
+
+            ProcessReads();
+            ProcessWrites();
+
+        }
+        private async void ProcessReads()
+        {
+            try
+            {
+                var args = new SocketAsyncEventArgs();
+                SemaphoreSlim pending = new SemaphoreSlim(0, 1);
+                args.Completed += AsyncCompleted;
+                args.UserToken = pending;
+
+                while (true)
+                {
+                    // we need a buffer to read into
+                    var buffer = input.Alloc(2048);
+                    SetBuffer(buffer.Memory, args);
+
+                    if (socket.ReceiveAsync(args)) //  initiator calls ReceiveAsync
+                    {
+                        // wait async for the semaphore to be released by the callback
+                        await pending.WaitAsync();
+                    }
+                    else
+                    {
+                        // if ReceiveAsync returns sync, we have the conch - nothing to do - we already received
+                    }
+                    // either way, need to validate
+                    if (args.SocketError != SocketError.Success)
+                    {
+                        throw new SocketException((int)args.SocketError);
+                    }
+                    int len = args.BytesTransferred;
+                    if(len <= 0)
+                    {
+                        // end of the socket
+                        await buffer.FlushAsync();
+                        break;
+                    }
+                    
+                    buffer.CommitBytes(len);
+                    await buffer.FlushAsync();
+                }
+                input.CompleteWriting();
+            }
+            catch (Exception ex)
+            {
+                input?.CompleteWriting(ex);
```

Comment:
it is a field; it can be null at any point it chooses - probably relating to disposal